addons/mail/static/src/models/welcome_view/welcome_view.js

```diff
@@ -0,0 +1,163 @@
+/** @odoo-module **/
+
+import { attr, one2one } from '@mail/model/model_field';
+import { create } from '@mail/model/model_field_command';
+import { registerNewModel } from '@mail/model/model_core';
+
+function factory(dependencies) {
+
+    const getNextGuestNameInputId = (function() {
+        let id = 0;
+        return () => ++id;
+    })();
+
+    class WelcomeView extends dependencies['mail.model'] {
+
+        /**
+         * @override
+         */
+        _created() {
+            super._created();
+            // Bind necessary until OWL supports arrow function in handlers: https://github.com/odoo/owl/issues/876
+            this.onClickJoinButton = this.onClickJoinButton.bind(this);
+            this.onInputGuestNameInput = this.onInputGuestNameInput.bind(this);
+            this.onKeydownGuestNameInput = this.onKeydownGuestNameInput.bind(this);
+        }
+
+        //----------------------------------------------------------------------
+        // Public
+        //----------------------------------------------------------------------
+
+        /**
+         * @static
+         * @param {string} name 
+         */
+        static async updateGuestNameServerSide(name) {
+            await this.env.services.rpc({
+                route: '/mail/guest/update_name',
+                params: { name },
+            });
+        }
+
+        /**
+         * @param {MouseEvent} ev 
+         */
+        async onClickJoinButton(ev) {
+            this.updateGuestNameIfAnyThenJoinChannel();
+        }
+
+        /**
+         * Handles OWL update on this WelcomeView component.
+         */
+        onComponentUpdate() {
+            this._handleFocus();
+        }
+
+        /**
+         * @param {KeyboardEvent} ev 
+         */
+        onInputGuestNameInput(ev) {
+            this._updateGuestNameWithInputValue();
+        }
+
+        /**
+         * @param {KeyboardEvent} ev 
+         */
+        onKeydownGuestNameInput(ev) {
+            if (ev.key === 'Enter') {
+                this.updateGuestNameIfAnyThenJoinChannel();
+            }
+        }
+
+        /**
+         * Redirects users to the related channel. If the current user is a
+         * guest, first updates their username with the value of guestNameInput.
+         */
+        async updateGuestNameIfAnyThenJoinChannel() {
+            if (this.messaging.currentGuest) {
+                await this.messaging.models['mail.welcome_view'].updateGuestNameServerSide(this.guestName.trim());
+            }
+            window.location = `/discuss/channel/${this.channelId}`;
```

Comment:
Ideally should be done with env.services or browser instead of window to be able to test it without actually changing the page.